### PR TITLE
Modify Span status according to specification

### PR DESCRIPTION
**Pull request by tigrannajaryan**
1. Renamed Status.code field to Status.deprecated_code.
2. Renamed StatusCode enum to Status.DeprecatedStatusCode.
3. Introduce new Status.code with corresponding StatusCode enum.
4. Added guidelines about how to use the fields to ensure backward compatibility.

### CODE REVIEW

File: opentelemetry/proto/trace/v1/trace.proto
Comment (bogdandrutu) [inline opentelemetry/proto/trace/v1/trace.proto:266]: Comments for the values?

File: opentelemetry/proto/trace/v1/trace.proto
Comment (bogdandrutu) [inline opentelemetry/proto/trace/v1/trace.proto:268]: Can this be make 1? I am confident that if we add more values they will be errors not ok so they will be grouped

**Review by bogdandrutu — APPROVED** [review]

**Review by jkwatson — APPROVED** [review]
Looks great!

File: opentelemetry/proto/trace/v1/trace.proto
Comment (tigrannajaryan) [reply opentelemetry/proto/trace/v1/trace.proto:268]: Yeah, I was considering that too. Changing.

File: opentelemetry/proto/trace/v1/trace.proto
Comment (tigrannajaryan) [reply opentelemetry/proto/trace/v1/trace.proto:267]: Added.

File: opentelemetry/proto/trace/v1/trace.proto
Comment (tigrannajaryan) [reply opentelemetry/proto/trace/v1/trace.proto]: Done.

**Review by iNikem — APPROVED** [review]

**Review by bryannaegele — COMMENTED** [review]
How does this avoid being a breaking change of a stable API when the name of the field and its values are being changed?

**Comment by Oberon00** [comment]
Also, since JSON is allowed, renaming is not even wire-compatible.

**Comment by bogdandrutu** [comment]
JSON is not stable

File: opentelemetry/proto/trace/v1/trace.proto
Comment (anuraaga) [inline opentelemetry/proto/trace/v1/trace.proto:256]: Not related to this PR but just noticed since it's defined very clearly here with a date - 12 months seems like a pretty short time for API stability post-GA.

**Review by anuraaga — APPROVED** [review]

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:280]:
```suggestion
  // To ensure any pair of senders and receivers continues to correctly signal and
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:281]:
```suggestion
  // interpret erroneous situations, the senders and receivers SHOULD follow these rules:
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:283]:
```suggestion
  // 1. Old senders and receivers that are not aware of the `code` field will continue using
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:284]:
```suggestion
  // the `deprecated_code` field to signal and interpret erroneous situation.
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:286]:
```suggestion
  // 2. New senders, which are aware of the `code` field SHOULD set both the `deprecated_code`
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:289]:
```suggestion
  //   if code==STATUS_CANONICAL_CODE_UNSET then `deprecated_code` MUST be
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:292]:
```suggestion
  //   if code==STATUS_CANONICAL_CODE_ERROR then `deprecated_code` MUST be
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:295]:
```suggestion
  //   if code==STATUS_CANONICAL_CODE_OK then `deprecated_code` MUST be
```

You could move that before the error rule to match the sorting order of the definition above.

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:300]:
```suggestion
  // 3. New receivers should look at both the `code` and `deprecated_code` fields in order
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:313]:
```suggestion
  //   ignored, the `code` field is the sole carrier of the status.
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:223]:
```suggestion
  // An optional final status for this span. Semantically when Status isn't set, it means
```

File: opentelemetry/proto/trace/v1/trace.proto
Comment (arminru) [inline opentelemetry/proto/trace/v1/trace.proto:254]:
```suggestion
  // This field is deprecated and is replaced by the `code` field below. See backward
```

**Review by arminru — DISMISSED** [review]

**Comment by tigrannajaryan** [comment]
> How does this avoid being a breaking change of a stable API when the name of the field and its values are being changed?

Names of the fields are not exposed OpenTelemetry API. Compatibility from protocol perspective means wire compatibility. names of the fields do not exist on the wire level.

We could keep the name of the field and of the enum values if there is a big desire to keep compatibility on the build level. I am not sure it is worth it and it was not part of the compatibility promise in the past (Collector codebase hides these details on purpose so that usage of OTLP protobuf messages in the codebase is not affected).

I am open to re-considering this and it does not affect the logic.

File: opentelemetry/proto/trace/v1/trace.proto
Comment (tigrannajaryan) [reply opentelemetry/proto/trace/v1/trace.proto:256]: 12 months is our formal definition of "stable": https://github.com/open-telemetry/community/blob/47813530864b9fe5a5146f466a58bd2bb94edc72/maturity-matrix.yaml#L57

**Comment by Oberon00** [comment]
> Names of the fields are not exposed in OpenTelemetry API. Compatibility from protocol perspective means wire compatibility. Names of the fields do not exist on the wire level.

What about OTLP/JSON?

**Comment by tigrannajaryan** [comment]
> > Names of the fields are not exposed in OpenTelemetry API. Compatibility from protocol perspective means wire compatibility. Names of the fields do not exist on the wire level.
> 
> What about OTLP/JSON?

Yes, in OTLP/JSON the Protobuf field names are directly exposed on the wire. However, OTLP/JSON is currently in "Alpha", we are free to change it:
image

File: opentelemetry/proto/trace/v1/trace.proto
Comment (tigrannajaryan) [reply opentelemetry/proto/trace/v1/trace.proto]: Done, thanks.

**Comment by bryannaegele** [comment]
@tigrannajaryan tooling does rely on these names, so it is a breaking change for every library that relies on the proto names to have to update. When these files (Common and Trace) were officially marked stable in August, I assumed that included stable for tooling, as well.

What I'm not clear on is why this enum is now being deprecated at all, nor why it's being marked as deprecated by prefixing the word to everything. The comment directly above this code states `The Status type defines a logical error model that is suitable for different programming environments, including REST APIs and RPC APIs.` but now those codes are marked deprecated and not replaced by anything equivalent.

It's unclear to me in https://github.com/open-telemetry/oteps/blob/master/text/trace/0136-error_flagging.md#status-mapping-schema if these codes are actually intended to be removed at this time. 

> As part of the specification, OpenTelemetry provides a mapping of semantic conventions to status codes. This removes any ambiguity as to what OpenTelemetry ships with out of the box.

What I did get from the OTEP was the addition of a code field `Status.Code` which has been named `StatusCanonicalCode in here and the addition of a `Status.Source` field which seems to not be in the PR.

I could be missing something as it's difficult to keep up. If so, I'd love to have more context on this PR. I greatly welcome the additions but I'm unclear on the motivation of the deprecation.

File: opentelemetry/proto/trace/v1/trace.proto
Comment (bogdandrutu) [reply opentelemetry/proto/trace/v1/trace.proto:256]: I think we need a new stability called "GA", that has stronger requirements.

**Comment by tigrannajaryan** [comment]
>It's unclear to me in https://github.com/open-telemetry/oteps/blob/master/text/trace/0136-error_flagging.md#status-mapping-schema if these codes are actually intended to be removed at this time.

@bryannaegele the OTEP's language may not be clear but the specification is clear: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#status

There is only one code - the StatusCanonicalCode, there is no other code in the spec that we need to also support in the protocol.

There is certainly no intent to have 2 different codes in the API or in the protocol. The OTEP's motivation was that the old enum did not make sense. This has been discussed in multiple meetings and in the OTEP.

>tooling does rely on these names, so it is a breaking change for every library that relies on the proto names to have to update. When these files (Common and Trace) were officially marked stable in August, I assumed that included stable for tooling, as well.

I am not strongly opposed to keeping the old field name and old enum name and value names. The reason I used "deprecated_" as a prefix was to discourage its usage. 

If there is no strong oposition I am happy to modify this PR to restore the names of the deprecated field and use new names for the new field and enum.

**Comment by bryannaegele** [comment]
@tigrannajaryan That doc changed 7 days ago where the API spec clearly stated that these status codes were to be used and thus available to other uses, such as messaging https://github.com/open-telemetry/opentelemetry-specification/blame/3e66e6fcc3bb8070db7ccbdd30005ae9f35ac29c/specification/trace/semantic_conventions/http.md#L40-L63. So the ability to set the status with these semantic conventions were in the API spec and does introduce a breaking change.

These status codes being set are also still in the referenced document for HTTP clients https://github.com/open-telemetry/opentelemetry-specification/blame/master/specification/trace/semantic_conventions/http.md#L109-L116.

**Comment by bryannaegele** [comment]
> If there is no strong oposition I am happy to modify this PR to restore the names of the deprecated field and use new names for the new field and enum.

If this change must be done for some reason, I'm simply asking what is the replacement for marking statuses on a span beyond `error`? Are users meant to devise their own schemes going forward?

**Comment by bryannaegele** [comment]
I just caught up and found https://github.com/open-telemetry/opentelemetry-specification/pull/966/files from 4 days ago, so this conversation is apparently moot and we'll just have to deal with the breaking changes.

**Comment by tigrannajaryan** [comment]
> I'm simply asking what is the replacement for marking statuses on a span beyond error?

The updated spec does not define anything for this. These has been discussed extensively in the spec SIG and error SIG. This PR is a consequence, it is not something new. This simply makes the corresponding change in the protocol.

>we'll just have to deal with the breaking changes.
Again, I am happy to keep the old field names in the protocol, but language SDK's have to deal with the spec change anyway. OTLP exporters in SDK will also have to comply with Compatibility requirements for senders that I added in this PR before GA. I don't think we want to release an SDK that has outdated understanding of Span Status.

**Comment by bryannaegele** [comment]
We can't all be in every SIG meeting :)

We'll deal with it. As you noted, we're going to have to remove all this from the SDK anyhow, so it's better to just get the pain over with.

I'm just looking for guidance at this point on how to guide users on how to substitute the functionality. For users that are already leveraging this functionality, what is the upgrade path for them? Are we just saying "Hey, error status codes aren't a thing anymore so you'll have to figure something out" or is there guidance/patterns we can provide? Asking as a user _and_ a maintainer 😆

**Comment by Oberon00** [comment]
Maybe we should add a compatibility semantic convention to ease transition at least?

**Comment by tigrannajaryan** [comment]
> I'm just looking for guidance at this point on how to guide users on how to substitute the functionality. For users that are already leveraging this functionality, what is the upgrade path for them? Are we just saying "Hey, error status codes aren't a thing anymore so you'll have to figure something out" or is there guidance/patterns we can provide? Asking as a user _and_ a maintainer 😆

@bryannaegele best to ask this in spec repo or gitter where it will get more eyeballs.

**Comment by tigrannajaryan** [comment]
> Maybe we should add a compatibility semantic convention to ease transition at least?

Can you please clarify what do you mean by this?

(I was about to merge this PR, just noticed you comment, so will hold for now).

**Comment by tigrannajaryan** [comment]
> Maybe we should add a compatibility semantic convention to ease transition at least?

@Oberon00  Can you please clarify what do you mean by this?

(I was about to merge this PR, just noticed you comment, so will hold for now).

**Comment by tigrannajaryan** [comment]
I have put a bit more thought in to this and I believe the transition in codebases will be easier to do if we keep the names of the deprecated field and enum unchanged. I am going to revert that part of the change. @bryannaegele FYI.

I am going to also wait for https://github.com/open-telemetry/opentelemetry-specification/issues/1069 resolution and do it in one PR. There is no point in introducing the field with wrong names and risk someone start using it and then immediately break it.

**Comment by tigrannajaryan** [comment]
I update this PR and removed "canonical" adjective to be inline with https://github.com/open-telemetry/opentelemetry-specification/pull/1081

I am using "code value" since "code" is unfortunately already used by a deprecated field which have to keep around for now.

**Comment by tigrannajaryan** [comment]
We have the final version of the filed in the spec: `StatusCode`. Unfortunately this is exactly the name of the existing enum which we cannot use. I am forced to keep what I invented - CodeValue. If there are better suggestions for the field and enum names please speak.
I am removing the WIP label, this is ready for final review.

**Comment by tigrannajaryan** [comment]
I would like to know what others think about the field names. The new name looks awkward to me. I would rather prefer to rename the old fields and use `StatusCode` for the new field/enum.

I am going to open an unrelated PR which legitimizes field renaming as soon as it does not break wire compatibility (which is fine up until we say OTLP/JSON is stable).

**Comment by tigrannajaryan** [comment]
FYI, regarding field name guarantees: https://github.com/open-telemetry/opentelemetry-proto/pull/225

**Comment by tigrannajaryan** [comment]
Given that https://github.com/open-telemetry/opentelemetry-proto/pull/225 got several approvals and is about to be merged I am inclined more and more to rename the fields since we no longer are limited by the compatibility guarantees. I am going to that unless I hear objections.

File: opentelemetry/proto/trace/v1/trace.proto
Comment (Oberon00) [inline opentelemetry/proto/trace/v1/trace.proto:278]: Having this floating around at the end looks odd. Maybe move it to the docs of StatusCode? Or at least mention "see compatibility notes below" for both code and deprecated_code.

File: opentelemetry/proto/trace/v1/trace.proto
Comment (Oberon00) [inline opentelemetry/proto/trace/v1/trace.proto:281]: >  SHOULD follow these rules

Either this should be a MUST or we need to describe what other alternatives are valid.

Comment (Oberon00) [reply opentelemetry/proto/trace/v1/trace.proto:281]: Also you use MUST within this should, which is odd (the other way round would be more understandable to me)

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-approvers @open-telemetry/specs-trace-approvers please review. I made a few changes back and forth in this PR. The PR description now reflects the latest state.

File: opentelemetry/proto/trace/v1/trace.proto
Comment (tigrannajaryan) [reply opentelemetry/proto/trace/v1/trace.proto]: Moved to top of Status.

File: opentelemetry/proto/trace/v1/trace.proto
Comment (tigrannajaryan) [reply opentelemetry/proto/trace/v1/trace.proto]: Changed to MUST.

**Review by Oberon00 — APPROVED** [review]

**Review by bogdandrutu — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-approvers @open-telemetry/specs-trace-approvers please review. Implementation in the Collector cannot proceed until this is merged. The Collector is currently non-compliant with the spec because of this.

**Comment by Oberon00** [comment]
I think there are already enough approvals, aren't there?

**Comment by tigrannajaryan** [comment]
> I think there are already enough approvals, aren't there?

@Oberon00 I re-requested reviews. Approvers previously saw and approved a different version of this PR and I think differences are pretty significant, so it is important for them to confirm they are still OK.

**Comment by Oberon00** [comment]
@tigrannajaryan Then you should dismiss the reviews in question to make this clearer.

**Comment by tigrannajaryan** [comment]
@arminru @anuraaga @jkwatson PTAL, I need to move forward, this is blocking Collector work. If I don't see any objections I will merge this today.

**Review by jkwatson — APPROVED** [review]

**Pull request merged**
